File: apps/node/src/contracts/onScreenLogTemplate.ts
/** Closed vocabulary and delimiter grammar, mirrored by Android OnScreenLogTemplate. */
export const onScreenLogTemplateNames = [
  "foreground_app.icon", "foreground_app.package_name", "foreground_app.version_code",
  "foreground_app.version_name", "device.manufacturer", "device.model",
  "system.language_code", "system.language_tag", "system.language_name",
] as const;

export function validateOnScreenLogTemplate(template: string): string | undefined {
  for (let index = 0; index < template.length;) {
    // Quadrupled delimiters are literal double delimiters, on either side.
    if (template.startsWith("{{{{", index) || template.startsWith("}}}}", index)) {
      index += 4;
    } else if (template.startsWith("{{", index)) {
      const end = template.indexOf("}}", index + 2);
      const name = end < 0 ? "" : template.slice(index + 2, end);
      if (!(onScreenLogTemplateNames as readonly string[]).includes(name)) return invalidTemplate();
      index = end + 2;
    } else if (template.startsWith("}}", index)) {
      return invalidTemplate();
    } else {
      index += 1;
    }
  }
  return undefined;
}

function invalidTemplate(): string {
  return `Invalid template placeholder; supported names: ${onScreenLogTemplateNames.join(", ")}. Escape delimiters as {{{{ and }}}}.`;
}

File: apps/android/shared/data/task/src/main/kotlin/clawperator/task/runner/OnScreenLogTemplate.kt
package clawperator.task.runner

/** Pure bounded grammar shared with the Node validator. Values are never parsed recursively. */
object OnScreenLogTemplate {
    const val MAX_EXPANDED_LENGTH = 8192
    const val MAX_VALUE_LENGTH = 256
    val names = setOf(
        "foreground_app.icon", "foreground_app.package_name", "foreground_app.version_code",
        "foreground_app.version_name", "device.manufacturer", "device.model",
        "system.language_code", "system.language_tag", "system.language_name",
    )

    data class Part(val literal: String? = null, val name: String? = null)

    fun parse(template: String): List<Part> {
        val parts = mutableListOf<Part>()
        val literal = StringBuilder()
        fun flush() {
            if (literal.isNotEmpty()) {
                parts.add(Part(literal = literal.toString()))
                literal.setLength(0)
            }
        }
        var index = 0
        while (index < template.length) {
            when {
                template.startsWith("{{{{", index) || template.startsWith("}}}}", index) -> {
                    literal.append(template.substring(index, index + 2))
                    index += 4
                }
                template.startsWith("{{", index) -> {
                    val end = template.indexOf("}}", index + 2)
                    val name = if (end < 0) "" else template.substring(index + 2, end)
                    if (name !in names) invalid()
                    flush()
                    parts.add(Part(name = name))
                    index = end + 2
                }
                template.startsWith("}}", index) -> invalid()
                else -> literal.append(template[index++])
            }
        }
        flush()
        return parts
    }

    private fun invalid(): Nothing = throw OnScreenLogValidationException(
        "Invalid template placeholder; supported names: ${names.joinToString(", ")}. Escape delimiters as {{{{ and }}}}.",
    )
}
